### PR TITLE
Implement outgoing link URL replacements.

### LINKED ISSUE #4078: Variable substitution in href of <a> tags?
I'd love to see the ability to do at least simple variable substitution in the `href` of an `<a>` tag.  specifically if I could get `CLIENT_ID` transmitted it would let me track sessions between the a cdn and regular site.  One of the big gripes we're getting from publishers is they have no idea if AMP is generating more or longer sessions or if users are just swiping in the carousel to another pubs content.  

I realize that some of the variables are real time so that would involve intercepting clicks and that bring it's own issues.  However  if we could just get the static stuff that would still be a huge win.

### PR DESCRIPTION
**Pull request by cramforce**
- Only exposes `QUERY_PARAM` and `CLIENT_ID`
- Requires opt-in per `<a>` tag.
- Only done for destinations going to the page's source or canonical origin.

Fixes #4078

### CODE REVIEW

**Comment by cramforce** [comment]
@zhouyx please take a look
CC @rudygalfi I went ahead and implemented this :)

**Comment by zhouyx** [comment]
One thing I don't understand.  We still do lazy initialization. What I see here is that `this.setAsync_('CLIENT_ID...` here is called no earlier than `this.set_('CLIENT_ID...`. Would this work as expected?

**Comment by cramforce** [comment]
@zhouyx Could you rephrase your question? I'm not sure I understand.

This is intended to only provide a CLIENT_ID into a link if there was another mechanism that previously requested it.

**Comment by zhouyx** [comment]
I see, I thought we would need to wait and replace the URL in all cases. 
So we are not replacing `CLIENT_ID` if we don't have it at that moment, would it be normal that url replacement will fail because we never wait for anything?

**Comment by cramforce** [comment]
Replacement happens on click, while the other mechanisms (analytics) load on page view, so they will probably have run by the time of the click.

File: src/service/url-replacements-impl.js [hunk 1]
```diff
@@ -213,8 +221,22 @@ export class UrlReplacements {
...
+        return cid;
       });
     });
+    // Synchronous alternative. Only works for scopes that were previously
+    // requested using the async method.
+    this.set_('CLIENT_ID', scope => {
+      if (!clientIds) {
+        return null;
+      }
+      return clientIds[dev().assertString(scope)];
```
Comment (zhouyx) [inline src/service/url-replacements-impl.js]: don't quite understand what's the scope here. Is it always the same as what we set `clientIds[scope` above? 
Would it be possible to get`clientIds[scope] = null` here?

File: src/service/url-replacements-impl.js [hunk 2]
```diff
@@ -606,6 +628,56 @@ export class UrlReplacements {
...
+    const url = parseUrl(href);
+    if (url.origin != parseUrl(docInfo.canonicalUrl).origin &&
+        url.origin != parseUrl(docInfo.sourceUrl).origin) {
+      user().warn('URL', 'Ignoring link replacement', href,
+          ' because the link does not go to the document\'s' +
+          ' source or canonical origin.');
+      return;
+    }
+    if (element[ORIGINAL_HREF_PROPERTY] == null) {
+      element[ORIGINAL_HREF_PROPERTY] = href;
```
Comment (zhouyx) [inline src/service/url-replacements-impl.js]: where is this used?

File: src/service/url-replacements-impl.js (hunk 1 above)
Comment (cramforce) [inline src/service/url-replacements-impl.js]: scope is the param passed of `CLIENT_ID(someScope)`. If the scope is the same, the client ID is the same, which should be what is achieved here. The return value of this statement is `undefined` if no async request for the client id was made.

File: src/service/url-replacements-impl.js (hunk 2 above)
Comment (cramforce) [inline src/service/url-replacements-impl.js]: Line 647. I'll add a comment as to what this is for.

**Comment by zhouyx** [comment]
lgtm

File: src/service/url-replacements-impl.js [hunk 3]
```diff
@@ -213,8 +221,22 @@ export class UrlReplacements {
           scope: dev().assertString(scope),
           createCookieIfNotPresent: true,
         }, consent);
+      }).then(cid => {
+        if (!clientIds) {
```
Comment (jridgewell) [inline src/service/url-replacements-impl.js]: Why not just initialize it above? It's just one object.

File: src/service/url-replacements-impl.js [hunk 4]
```diff
@@ -606,6 +628,59 @@ export class UrlReplacements {
...
+      return;
+    }
+    if (element[ORIGINAL_HREF_PROPERTY] == null) {
+      element[ORIGINAL_HREF_PROPERTY] = href;
+    }
+    const supportedReplacements = {
+      'CLIENT_ID': true,
+      'QUERY_PARAM': true,
+    };
+    const requestedReplacements = {};
```
Comment (jridgewell) [inline src/service/url-replacements-impl.js]: We could reuse the same `supportedReplacements` by settings its values to `false`.

Comment (cramforce) [inline src/service/url-replacements-impl.js]: Yeah, it also isn't actually used like that, i think. I don't care about allocation on click :)
